Gem/Code/Source/Components/ScriptableDecalComponent.cpp

```diff
@@ -0,0 +1,164 @@
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                    ->Attribute(AZ::Edit::Attributes::Category, "Graphics")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZStd::vector<AZ::Crc32>({ AZ_CRC_CE("Level") }))
+                    ;
+            }
+        }
+
+        AZ::BehaviorContext* behaviorContext = azrtti_cast<AZ::BehaviorContext*>(context);
+        if (behaviorContext)
+        {
+            behaviorContext->EBus<DecalRequestBus>("RequestBus")
+                ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Common)
+                ->Attribute(AZ::Script::Attributes::Category, "Rendering")
+                ->Attribute(AZ::Script::Attributes::Module, "rendering")
+                ->Attribute(AZ::Script::Attributes::Scope, AZ::Script::Attributes::ScopeFlags::Common)
+                ->Event(
+                    "SpawnDecal",
+                    &DecalRequestBus::Events::SpawnDecal)
+                ;
+        }
+    }
+
+    void ScriptableDecalComponent::GetProvidedServices(AZ::ComponentDescriptor::DependencyArrayType& services)
+    {
+        services.push_back(AZ_CRC_CE("ScriptableDecalService"));
+    }
+
+    void ScriptableDecalComponent::GetIncompatibleServices(AZ::ComponentDescriptor::DependencyArrayType& services)
+    {
+        services.push_back(AZ_CRC_CE("ScriptableDecalService"));
+    }
+
+    void ScriptableDecalComponent::Activate()
+    {
+        AZ_Printf("ScriptableDecalComponent", "Activate");
+        m_decalFeatureProcessor = AZ::RPI::Scene::GetFeatureProcessorForEntity<AZ::Render::DecalFeatureProcessorInterface>(GetEntityId());
+        if (m_decalFeatureProcessor)
+        {
+            AZ::RPI::Scene* scene = m_decalFeatureProcessor->GetParentScene();
+            DecalRequestBus::Handler::BusConnect(scene->GetId());
+            AZ::TickBus::Handler::BusConnect();
+        }
+    }
+
+    void ScriptableDecalComponent::Deactivate()
+    {
+        AZ::TickBus::Handler::BusDisconnect();
+        DecalRequestBus::Handler::BusDisconnect();
+        m_decalFeatureProcessor = nullptr;
+    }
+
+    void ScriptableDecalComponent::SpawnDecal(const AZ::Transform& worldTm, AZ::Data::AssetId materialAssetId, const SpawnDecalConfig& config)
+    {
+        DecalHandle handle = m_decalFeatureProcessor->AcquireDecal();
+
+        AZ::Transform transform = worldTm * AZ::Transform::CreateUniformScale(config.m_scale);
+
+        m_decalFeatureProcessor->SetDecalTransform(handle, transform);
+        m_decalFeatureProcessor->SetDecalMaterial(handle, materialAssetId);
+        m_decalFeatureProcessor->SetDecalOpacity(handle, config.m_opacity);
+        m_decalFeatureProcessor->SetDecalAttenuationAngle(handle, config.m_attenutationAngle);
+        m_decalFeatureProcessor->SetDecalSortKey(handle, config.m_sortKey);
+
+        uint32_t currentTimeMs = GetCurrentTimeMs();
```

Comment:
I'm surprised this compiled, doesn't GetCurrentTimeMs() return an AZ::TimeMs type?

Reply:
That's just a local function in `ScriptableDecalComponent()` In an earlier version of this file I needed that code in a couple places so I put it in a function, but that's no longer the case so I think I can just remove it.